File: benchmarks-QuantumOptics.jl/multiplication_sparse_sparse.jl
using QuantumOptics
using BenchmarkTools
using SparseArrays
include("benchmarkutils.jl")

using Random; Random.seed!(0)

basename = "multiplication_sparse_sparse"

samples = 2
evals = 5
cutoffs = [50:50:1000;]
S = [0.1, 0.01, 0.001]
Nrand = 5

function setup(N, s)
    op1 = sprand(ComplexF64, N, N, s)
    op2 = sprand(ComplexF64, N, N, s)
    op1, op2
end

function f(op1, op2)
    op1*op2
end

for s in S
    name = basename * "_" * replace(string(s), "." => "")
    println("Benchmarking: ", name)
    print("Cutoff: ")
    results = []
    for N in cutoffs
        print(N, " ")
        T = 0.
        for i=1:Nrand
            op1, op2 = setup(N, s)
            T += @belapsed f($op1, $op2) samples=samples evals=evals
        end
        push!(results, Dict("N"=>N, "t"=>T/Nrand))
    end
    println()
    benchmarkutils.save(name, results)
end

File: benchmarks-QuTiP/multiplication_sparse_sparse.py
import scipy.sparse as sp
import benchmarkutils

name = "multiplication_sparse_sparse"

samples = 5
evals = 100
cutoffs = range(50, 1001, 50)

def setup(N):
    s = 0.01
    op1 = sp.rand(N, N, s, dtype=float)*0.2j
    op2 = sp.rand(N, N, s, dtype=float)*0.1j
    return op1, op2

def f(op1, op2):
    return op1*op2

print("Benchmarking:", name)
print("Cutoff: ", end="", flush=True)
results = []
for N in cutoffs:
    print(N, "", end="", flush=True)
    op1, op2 = setup(N)
    t = benchmarkutils.run_benchmark(f, op1, op2, samples=samples, evals=evals)
    results.append({"N": N, "t": t})
print()

benchmarkutils.save(name, results)
